contracts-bedrock: add Kontrol CI setup

JuanCoRo opened the pull request:
<!-- Contributions welcome! See https://github.com/ethereum-optimism/.github/blob/master/CONTRIBUTING.md -->

**Description**

This PR adds CI support for running Kontrol proofs. In particular:
- Creates a new directory `tests/kontrol` where all things Kontrol live
- Adds a dummy proof under `tests/kontrol/Counter.t.sol`
- Sets up CircleCI to trigger the compilation and execution of such proof when changes to the `test/kontrol` directory or `OptimismPortal.sol` file are made

**Tests**

The test added is the above-mentioned proof itself.

**Additional context**

This PR is the first step to integrate more developed Kontrol proofs into CI.

JuanCoRo committed:
Add dummy verification

JuanCoRo committed:
Add logging to execution script

JuanCoRo committed:
Add `Counter.sol` file to `bedrock/src/L1` folder

JuanCoRo committed:
Export `FOUNDRY_PROFILE` variable

F-WRunTime committed:
Updatin circleci config to run kontrol tests when changes occur in kontrol folder, top-level

JuanCoRo committed:
Update packages/contracts-bedrock/src/L1/Counter.sol

JuanCoRo committed:
Update packages/contracts-bedrock/test/kontrol/Counter.t.sol

JuanCoRo committed:
Update packages/contracts-bedrock/test/kontrol/Counter.t.sol

JuanCoRo committed:
Update packages/contracts-bedrock/test/kontrol/Counter.t.sol

JuanCoRo committed:
Update .circleci/config.yml

JuanCoRo committed:
Move `Counter.sol` to Kontrol folder

tynes commented on packages/contracts-bedrock/foundry.toml:
```diff
@@ -59,3 +59,21 @@ runs = 512
+out = 'kout'
+test = 'test/kontrol'
+script = 'scripts-kontrol'
+remappings = [
```
Why do we need to redefine the remappings here? Does it not inherit from the default profile or do they need to be different? Would appreciate a comment inline

JuanCoRo replied:
Hey, you're absolutely correct. No need to redefine remappings there. Has been removed now. Initially, it contained slightly different remappings, adding the [kontrol-cheatcodes](https://github.com/runtimeverification/kontrol-cheatcodes/) library, and I copy-pasted just blindly with the redundancy you mentioned. But for this demo, that library is not needed

JuanCoRo committed:
Remove redundant remappings

JuanCoRo committed:
Change `when` key in favor of `check-changed`

mds1 commented on .circleci/config.yml:
```diff
@@ -1269,6 +1269,46 @@ jobs:
+      - run:
+          name: "Start Docker Container"
+          command: |
+            export KONTROL_VERSION=$(cat deps/kontrol_release)
```
@tynes What's the preferred way to specify a kontrol version here? Looks like creating a `.kontrolrc` file at the repo root is the way to go, but wanted to confirm

tynes replied:
We currently use the dotfile approach like you mentioned. Its growing a ton with the number of tools but we should prefer to follow it

JuanCoRo replied:
Great, just added `.kontrolrc` at the root directory of the repo

mds1 commented on .circleci/config.yml:
```diff
@@ -1269,6 +1269,46 @@ jobs:
+          name: 'Run Kontrol'
+          command: |
+            docker exec -u user optimism-ci bash -c 'cd packages/contracts-bedrock && ./test/kontrol/kontrol/run-kontrol.sh'
+      - run:
+          name: Create Results Tar
+          command: |
+            tar -czvf results.tar.gz kout/proofs
+      - store_artifacts:
+          path: results.tar.gz
+          destination: results
+          when: always
```
The proofs currently can take a few hours to run, so we upload the proof artifacts so CI, this way if a run fails we can download the artifacts and start debugging immediately instead of re-running the proofs locally.

The artifact files would be about 50-100MB in size, and we have CircleCI set to persist artifacts for 30 days. I want to confirm if this configuration is ok? @tynes @mslipper 

Once we get proof times down, we can consider removing this (but even still they'll likely take 20-30mins so it might be nice to keep it as-is)

tynes replied:
It would be nice to have a cost estimation of this. I looked and didn't seem to find a pricing model for artifacts.

mds1 replied:
From https://app.circleci.com/settings/plan/github/ethereum-optimism/overview it says our plan gives us 200 GB of storage. And https://app.circleci.com/settings/plan/github/ethereum-optimism/usage says we're currently at 46.3 GB (denominated in "storage-months" which I'm guessing is related to the fact that by default artifacts last for 30 days). So as long as we stay under the 200GB limit I don't think it should add any cost

mslipper replied:
How often do we plan on running this job? I'm more concerned about CCI compute credit usage than storage given that the job could take hours. Using the artifacts store is fine for now; we can change this to use a GCS bucket later if it becomes a problem.

mds1 replied:
It will only run on changes to affected contracts, which eventually will be the OptimismPortal, L1StandardBridge, L1ERC721Bridge, and the L1CrossDomainMessenger. For now we just have a dummy counter contract, then we'll incrementally add each of the above contracts. I think a good path forward here is:
- Switch to xlarge resource class to unblock this PR. These proofs will really only run once since it's a dummy contract for now.
- Monitor CI times as we start adding real proofs. If they become too long and/or too frequent, temporarily switch to a nightly CI job that creates an issue on failure
- Once kontrol has proof parallelization we can add proofs back into the standard PR flow and consider options for keeping credit usage low (such as offloading compute)

mds1 commented on packages/contracts-bedrock/foundry.toml:
```diff
@@ -59,3 +59,13 @@ runs = 512
+
+[profile.kontrol]
+src = 'src/L1'
+out = 'kout'
```
Let's add `kout` to the `.gitignore`

JuanCoRo replied:
Done :+1:

mds1 commented on packages/contracts-bedrock/test/kontrol/kontrol/run-kontrol.sh:
```diff
@@ -0,0 +1,79 @@
+export FOUNDRY_PROFILE=kontrol
+
+# Create a log file to store standard out and standard error
+LOG_FILE="run-kontrol-$(date +'%Y-%m-%d-%H-%M-%S').log"
```
I think it might be cleaner to put the log file in the `kout` or `kout-logs` directory to keep the root of this package a bit cleaner (if we do `kout-logs` we should make sure that's also in the `.gitignore`)

@tynes let me know if you have any preference here

JuanCoRo replied:
I went ahead and added the logging folder under `tests/kontrol/kontrol/logs`. The reason being we might not want `forge clean` to delete the logs created. Let me know if this is alright with you

mds1 replied:
Good call, I like that

mds1 commented on .circleci/config.yml:
```diff
@@ -1269,6 +1269,46 @@ jobs:
+      - checkout
+      - check-changed:
+          patterns: contracts-bedrock/test/kontrol,contracts-bedrock/src/L1/OptimismPortal\.sol
+      - run:
+          name: Checkout submodules
+          command: |
+            git submodule sync
+            git submodule update --init --recursive
```
We should be able to delete this, `forge build` will automatically install any missing dependencies when it's invoked
```suggestion
```

mds1 commented on .circleci/config.yml:
```diff
@@ -1269,6 +1269,46 @@ jobs:
+      - store_artifacts:
+          path: results.tar.gz
+          destination: results
+          when: always
```
`Property 'when' is not allowed`, based on the schema [here](https://github.com/SchemaStore/schemastore/blob/master/src/schemas/json/circleciconfig.json)

JuanCoRo replied:
What should it be substituted with? Or should it only be removed? I'm a bit lost here

mds1 replied:
From the [docs](https://circleci.com/docs/artifacts/) it seems like it can be deleted

JuanCoRo replied:
I've removed it. Following [this](https://circleci.com/docs/artifacts/#uploading-core-files), we might want to add `when: on_fail` after ensuring that step works correctly. But adding the `on_fail` flag is not necessary I would say

mds1 replied:
So currently the artifacts are always uploaded, not just on failure? If so I think that's ok so I'll resolve this thread

mds1 commented on .circleci/config.yml:
```diff
@@ -1269,6 +1269,46 @@ jobs:
             ./ops/scripts/ci-match-values-between-files.sh "<< parameters.file1_path >>" "<< parameters.pattern_file1 >>" "<< parameters.file2_path >>" "<< parameters.pattern_file2 >>"
+
+  test-proofs:
+    machine: true
```
Also seeing this warning
```
CircleCI advises against using "machine: true", as support for this feature is not guaranteed to continue in the future.
		
	You can replace it with the following explicit declaration, which uses the same image.
	machine:
		image: ubuntu-2204:2023.07.2
```

JuanCoRo replied:
I added
```yaml
  test-proofs:
      machine:
        image: ubuntu-2204:2023.07.2
```
But, by looking at the other images, I mostly see either `image: ubuntu-2204:2022.10.2` or `image: ubuntu-2204:2022.07.1`. Let me know if any of these is preferable

mds1 replied:
cc @tynes

tynes replied:
We should move to a `params` based approach to ensure that all of them are the same version and a single line diff can update all of the versions that are used, similar to `ci-builder`. This change is out of scope of this PR, no strong  opinions on which version to use

tynes replied:
https://github.com/ethereum-optimism/optimism/pull/8321

mds1 replied:
Thanks, so @JuanCoRo lets rebase this against the updated develop branch and use the same approach that was introduced in #8321

F-WRunTime replied:
Doing a test of trying to utilize this method appears may be a bit tricky as the intent is to use our published kontrol images and provide the ability to pull the latest version of that image from the .kontrolrc file and placing it into the 
`image: ` declaration tag. 
Maybe there's  a method to do this I'm not familiar with in circleci? 

Using 
```
parameters:
  ci_builder_image:
    type: string
    default: us-docker.pkg.dev/oplabs-tools-artifacts/images/ci-builder:v0.30.0
```
Would be nice but I'm not seeing a clear way to call this from within the job/docker: image and set what we want to run with.

mds1 replied:
@F-WRunTime I think this comment (https://github.com/ethereum-optimism/optimism/pull/8271#issuecomment-1832684513) is along the same lines as this thread, so moving the discussion here to keep things organized.

If interpolation doesn't work in that `default` key I'm not sure offhand if there's a way, but can look into it

JuanCoRo committed:
Add `kout` to .gitignore

JuanCoRo committed:
Add dedicated log folder for Kontrol

JuanCoRo committed:
Explicit declaration for `machine: true`

JuanCoRo committed:
Remove `Checkout submodules` in favor of `forge build`

`git submodule sync` and `git submodule update --init --recursive` should be subsumed by `forge build`, which is run when the execution script is triggered. Hence the removal of this step

Co-authored-by: Matt Solomon <matt@mattsolomon.dev>

JuanCoRo committed:
Format `Counter.t.sol` properly

JuanCoRo committed:
Merge remote-tracking branch 'origin/develop' into rv/feature-ci

JuanCoRo committed:
Remove `when: always` key

JuanCoRo committed:
Add `.kontrolrc` file

F-WRunTime commented on .circleci/config.yml:
```diff
@@ -1345,6 +1345,41 @@ jobs:
+            tar -czvf results.tar.gz kout/proofs
+      - store_artifacts:
+          path: results.tar.gz
+          destination: results
```
We can't drop the when clause here as it's critical to make sure we're publishing helpful pass/fail information. As I understand this saves local debug reruns from wasting time and allowing to jump back to the failure. @JuanCoRo is that accurate?

mds1 replied:
Yea that's accurate. But I think always uploading is also ok, because maybe a proof passed unexpectedly that we want to debug. My suggestion would be to leave it as "always upload" which should be simpler, and we can improve this in the future if needed

F-WRunTime commented on .circleci/config.yml:
```diff
@@ -1345,6 +1345,41 @@ jobs:
+            tar -czvf results.tar.gz kout/proofs
+      - store_artifacts:
+          path: results.tar.gz
+          destination: results
```
```suggestion
          destination: results
          when: "on_fail" 
```

F-WRunTime replied:
We need some sort of catch like the above proposed. BUT It seems circleCI is a bit more complicated for using conditionals on other commands such as `store_artifacts:`

I believe based on some research we'll need to write a special command that handles the on failure upload piece. Something like this: 

```YAML
commands:
  upload_artifact:
    description: On pass or fail, upload proof
    parameters:
      when:
        type: string

    steps:
      - run: 
          command: |
            echo "Creating Build Results Artifact"
            tar -czvf results.tar.gz kout/proofs
            echo "<<parameters.when>>"
          name: Upload Artifact on Pass/Fail
          when: always
      - store_artifacts:
          path: results.tar.gz
          destination: results

 test-proofs:
    machine:
      image: ubuntu-2204:2023.07.2
    steps:
      - checkout
      - check-changed:
          patterns: contracts-bedrock/test/kontrol,contracts-bedrock/src/L1/OptimismPortal\.sol
      - run:
          name: "Start Docker Container"
          command: |
            export KONTROL_VERSION=$(cat ../../.kontrolrc)
            docker run --name optimism-ci \
                    --rm \
                    -v $(pwd)/kout/proofs:/home/user/workspace/packages/contracts-bedrock/kout/proofs/ \
                    --interactive \
                    --tty \
                    --detach \
                    --user root \
                    --workdir /home/user/workspace \
                    runtimeverificationinc/kontrol:ubuntu-jammy-${KONTROL_VERSION}
            docker cp . optimism-ci:/home/user/workspace
            docker exec optimism-ci chown -R user:user /home/user
      - run:
          name: 'Run Kontrol'
          command: |
            docker exec -u user optimism-ci bash -c 'cd packages/contracts-bedrock && ./test/kontrol/kontrol/run-kontrol.sh'
      - upload_artifact:
        when: "always" 
```

This is at a cursory glance to see if ya'll have a suggestion on how you'd like to do this. I think the alternative would be to write a custom orb that handles this scenario in fewer lines of code to the CI config. This is likely a reasonable option though would take me a bit more time to figure out if you all haven't written one of these before. But I'd be happy to take a look. 

Reference: 
https://discuss.circleci.com/t/on-fail-for-commands-in-steps/42496

tynes commented on .circleci/config.yml:
```diff
@@ -1345,6 +1345,41 @@ jobs:
+      - run:
+          name: "Start Docker Container"
+          command: |
+            export KONTROL_VERSION=$(cat ../../.kontrolrc)
```
Can we move this logic into a shell script so its portable and possible to run locally more easily? there is a `scripts` directory in `contracts-bedrock`

JuanCoRo replied:
Cc @F-WRunTime

tynes commented on packages/contracts-bedrock/foundry.toml:
```diff
@@ -59,3 +59,13 @@ runs = 512
+################################################################
+
+[profile.kontrol]
+src = 'src/L1'
```
Does this mean it will never be able to operate on contracts in other directories?

JuanCoRo replied:
As long as using the `kontrol` profile, no. But this is intended since the only contracts Kontrol cares about are in that folder. Including other folders would only bloat the Kontrol building stage and add extra noise. The reasoning behind this restriction is only to build what we need, but let me know if you have any further concerns

mds1 replied:
Agreed this should be ok for now and I believe it will reduce runtime and artifact size since less is being compiled. Going to mark this resolved, since we can change this if we ever add proofs for other directories

F-WRunTime committed:
Update config.yml

F-WRunTime committed:
Moving run steps to a script

F-WRunTime committed:
Update permissions, Add comments to script

F-WRunTime committed:
Use Kontrol image for base image

F-WRunTime committed:
Typo in image name

F-WRunTime committed:
Bad end quote in script

F-WRunTime commented:
@tynes I have now a simpler setup running using the RV kontrol image published to dockerhub pulling it as part of the initial setup like so: 

```
  kontrol-tests:
    docker:
      - image: runtimeverificationinc/kontrol:ubuntu-jammy-0.1.74
    steps:
      - checkout
      - check-changed:
          patterns: contracts-bedrock/test/kontrol,contracts-bedrock/src/L1/OptimismPortal\.sol
      - run:
          name: Run Kontrol Tests
          command: |
            ./packages/contracts-bedrock/scripts/kontrol.sh           
```

Honestly can probably drop the script as it's a couple lines that may not be worth burying into the folder structure, but happy to leave it. 

What I don't love currently and would like to make more robust is the 
```
    docker:
      - image: runtimeverificationinc/kontrol:ubuntu-jammy-0.1.74
```
Is there a more robust way such that we could do something more like the following?
```
    docker:
      - image: runtimeverificationinc/kontrol:ubuntu-jammy-<<pipeline.parameters.kontrol_version>> 
```
Where that version is set by the .kontrolrc ? 

The intent here is we would like to be able to push versions of kontrol down when tests have first passed on our end within our forked repository and testing harnesses. There are other methods to update this tag version. I would like this to better match your current schema.

JuanCoRo commented:
I see the only failing job is `kontrol-tests`. We have a failure with a [kompile return code 137](https://app.circleci.com/pipelines/github/ethereum-optimism/optimism/34449/workflows/c293748b-2c0b-4abd-98c1-0224bd1b4f7e/jobs/1571935?invite=true#step-103-22432_19), meaning resource exhaustion. Most probably OOM. I've ran the job commands on the same docker image on my machine and it passes. However, the kompilation step will hog +10gb RAM. How many resources has `kontrol-tests` allocated?

mds1 commented on packages/contracts-bedrock/test/kontrol/kontrol/run-kontrol.sh:
```diff
@@ -0,0 +1,83 @@
+base_module=PAUSABILITY-LEMMAS
+module=CounterTest:${base_module}
+
+rekompile=--rekompile
+rekompile=
```
@JuanCoRo I'm supportive of adding these comments suggested by the bot so the script is more self-documenting. Just one comment at the top of the script explaining the usage/pattern seems sufficient

mds1 replied:
All suggested comments have been implemented in 4228833

JuanCoRo committed:
Set `kontrol-tests` resource class to `xlarge`

JuanCoRo committed:
`run-kontrol.sh`: Add description of setting-unsetting pattern

JuanCoRo committed:
Merge remote-tracking branch 'origin/develop' into rv/feature-ci

F-WRunTime committed:
Need to point to the proper directory where the test are run and the kout folder is generated relative to repository top level

F-WRunTime commented:
Alright we have a successful run publishing the files we need during a success. The path was a bad pointer relative to the top level repository. 
Success is [here](https://app.circleci.com/pipelines/github/ethereum-optimism/optimism/35300/workflows/1c22ebbb-62f8-43e1-b48f-95d8e2754889/jobs/1603962)

mds1 reviewed (approved):
LGTM, thank you!

tynes commented on .circleci/config.yml:
```diff
@@ -1340,6 +1340,27 @@ jobs:
+      - run:
+          name: Run Kontrol Tests
+          command: |
+            ./packages/contracts-bedrock/scripts/kontrol.sh           
```
Does this need to run in the context of the root of the repo? If no, we should run it in the context of the contracts package and wrap the call to the script in a pnpm command to follow suit of the other tests

tynes replied:
See
https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/scripts/generate-l2-genesis.sh#L10
On how to do this

F-WRunTime replied:
Sure thing the script was to address the ask for a script to wrap the steps taken to run the tool in the proper context. I can take a look at using pnpm instead.

tynes commented on packages/contracts-bedrock/scripts/kontrol.sh:
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env bash
+set -exuo pipefail
+# Setup Kontrol 
+cd packages/contracts-bedrock && ./test/kontrol/kontrol/run-kontrol.sh
```
This should be able to run in context of any directory, not just in context of the root of the repo

tynes replied:
https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/scripts/generate-l2-genesis.sh#L10

tynes reviewed (changes requested):
I don't buy we need this script

JuanCoRo committed:
Try `large` containers for `kontrol-tests`

JuanCoRo committed:
Back to xlarge

F-WRunTime committed:
Saving testable scripts and pnpm package.json additions

F-WRunTime committed:
Script section name change

F-WRunTime committed:
Fixing path, dirname stripped a necessary folder

F-WRunTime committed:
Updating job to install docker-cli to ci image for calling and executing the kontrol image from remote

F-WRunTime committed:
Remove version from script input, use kontrolrc file

F-WRunTime committed:
wrong path to kontrorc file. Remove version from circle config

F-WRunTime committed:
Removing old script, consolidated to run-kontrol.sh

F-WRunTime commented:
To hold on further runs on the CI pipeline. I have consolidated the run scripts however I'm still running into some minor issues in running the [forge step](https://app.circleci.com/pipelines/github/ethereum-optimism/optimism/35774/workflows/27ae2361-4fc7-407c-a627-d20256840ca0/jobs/1624266?invite=true#step-105-2033_7). Not sure is this related to the foundry problem we were discussing yesterday? 

There is also an issue still to make sure properly captures when a failure occurs. I am seeing failures with foundry as well testing on our cirlceCi instance.. So I may have to come back to this and get some expertise on the tests themselves.

F-WRunTime committed:
Fixing missing failure

F-WRunTime committed:
update loggint o publish results on pass AND FAIL

F-WRunTime committed:
Add back docker stop

F-WRunTime committed:
Missing an environment varibale for foundry profile in container

F-WRunTime committed:
Isolate docker cleanup function, retesting previous failure

F-WRunTime committed:
Recursively check out bedrock submodules

F-WRunTime commented on packages/contracts-bedrock/test/kontrol/kontrol/run-kontrol.sh:
```diff
@@ -0,0 +1,161 @@
+    runtimeverificationinc/kontrol:ubuntu-jammy-${KONTROL_RELEASE}
+
+    # Copy test content to container
+    docker cp --follow-link $WORKSPACE_DIR/. ${CONTAINER_NAME}:/home/user/workspace
```
Is this a version mismatch between AI knowledge base or what is run in CI? I have here follow-link is an available option 
```
DOCKER(1)                                                                                     Docker User Manuals                                                                                     DOCKER(1)

NAME
       docker-cp - Copy files/folders between a container and the local filesystem

SYNOPSIS
       docker cp [OPTIONS] CONTAINER:SRC_PATH DEST_PATH|-      docker cp [OPTIONS] SRC_PATH|- CONTAINER:DEST_PATH

DESCRIPTION
       Alias for docker container cp.

OPTIONS
       -a, --archive[=false]      Archive mode (copy all uid/gid information)

       -L, --follow-link[=false]      Always follow symbol link in SRC_PATH

       -h, --help[=false]      help for cp

       -q, --quiet[=false]      Suppress progress output during copy. Progress output is automatically suppressed if no terminal is attached
       
```

F-WRunTime committed:
Creating tar was not necessary, removing.

F-WRunTime commented:
I ended up having to add back the following to the .circleci/config.yml as it does not appear `forge build` is pulling the submodules in for `packages/contracts-bedrock/`
```
  kontrol-tests:
    docker:
      - image: << pipeline.parameters.ci_builder_image >>
    resource_class: xlarge
    steps:
      - checkout
      - run:
          name: Checkout Submodule
          command: |
            git submodule update --init --recursive
```
We were seeing a strange behavior that files did not exist on execution of the tools [here](https://app.circleci.com/pipelines/github/ethereum-optimism/optimism/35847/workflows/ad7acb88-9ac9-4f3d-96bc-dbdee7c0823d/jobs/1626529?invite=true#step-105-2018_97)

This is a reference to a submodule component where those files were supposed to be
```
 ecd2ca2cd7cac116f7a37d0e474bbb3d7d5e1c4d packages/contracts-bedrock/lib/openzeppelin-contracts (v4.7.3)
 ```
So I added it back and now we're [running successfully](https://app.circleci.com/pipelines/github/ethereum-optimism/optimism/35858/workflows/e94135e6-7ed8-496a-ad7b-5a07e7cb903a/jobs/1626912)

F-WRunTime committed:
Updating artifact upload reference path, and destination name

mds1 commented on packages/contracts-bedrock/test/kontrol/kontrol/run-kontrol.sh:
```diff
@@ -0,0 +1,161 @@
+#########################
+# kontrol build options #
+#########################
+# NOTE: This script has a recurring pattern of setting and unsetting variables, such as `rekompile`. Such a pattern is intended for easy use while locally developing and executing the proofs via this script. Comment/uncomment the empty assignment to activate/deactivate the corresponding flag
```
readability nit
```suggestion
# NOTE: This script has a recurring pattern of setting and unsetting variables,
# such as `rekompile`. Such a pattern is intended for easy use while locally
# developing and executing the proofs via this script. Comment/uncomment the
# empty assignment to activate/deactivate the corresponding flag
```

mds1 commented on packages/contracts-bedrock/test/kontrol/kontrol/run-kontrol.sh:
```diff
@@ -0,0 +1,161 @@
+#!/bin/bash
+set -u
```
```suggestion
set -euo pipefail
```

F-WRunTime replied:
I removed on error and pipefail for the purpose of enforcing use of the trap function to capture these errors with some additional cleanup, log generation, and results collection prior to the script quick failing. Is this acceptable?

mds1 replied:
Ah, yes that works for me, I missed that :)

F-WRunTime committed:
Create 2 output files from results, 1 with static name.

JuanCoRo committed:
Update packages/contracts-bedrock/test/kontrol/kontrol/run-kontrol.sh

Co-authored-by: Matt Solomon <matt@mattsolomon.dev>

mds1 commented on packages/contracts-bedrock/test/kontrol/kontrol/run-kontrol.sh:
```diff
@@ -0,0 +1,164 @@
+#!/bin/bash
+set -u
```
Resolving per https://github.com/ethereum-optimism/optimism/pull/8271#discussion_r1419698276

F-WRunTime committed:
Fixing shellcheck analysis

F-WRunTime committed:
Per https://github.com/foundry-rs/foundry/issues/6519, run update first

F-WRunTime committed:
Instll as well to run git submodule update

F-WRunTime committed:
Reverting to git direct method of checkout

F-WRunTime committed:
Step description

tynes commented on .circleci/config.yml:
```diff
@@ -1340,6 +1340,45 @@ jobs:
+      - run: 
+          name: Install Docker CLI
+          command: |
+            # Add Docker's official GPG key:
```
We should add this to the `ci-builder` image itself eventually

mds1 replied:
Now tracking in https://github.com/ethereum-optimism/optimism/issues/8529

tynes reviewed (approved):
removing my request for changes, generally looks good to me now

F-WRunTime committed:
Updating to fix most shellcheck findings. Kontrol commands do not run properly when the variable is quoted

F-WRunTime commented:
> removing my request for changes, generally looks good to me now

Doing a little clean up for shellcheck. Shellcheck is not going to be happy with [these kontrol functions](https://github.com/ethereum-optimism/optimism/pull/8271/files#diff-e4efc7eeaecf4c54c6fa0fdb4c77b654cf33c98208557352588e618fe50880ceR32-R56) as quoting them causes kontrol to parse the inputs incorrectly.

F-WRunTime committed:
Updating to fix most shellcheck findings. Kontrol commands do not run properly when the variable is quoted

F-WRunTime commented:
I can't explain [this ](https://app.circleci.com/pipelines/github/ethereum-optimism/optimism/36010/workflows/ffb19fea-6e98-45da-8f63-c6ce75c6f962/jobs/1634238?invite=true#step-108-159) failure. It does not appear related, can this fail intermittently? Tried to see if I could just rerun the  single test and doesn't appear I have permissions to do so.

JuanCoRo committed:
Merge remote-tracking branch 'origin/develop' into rv/feature-ci

JuanCoRo commented:
Merging the latest `develop` is how I've been fixing this issue. Which seems to have worked now as well
